### PR TITLE
DOC-674: k shortest path

### PR DESCRIPTION
**Pull request by diana-macrometa**
I broke up the K Shortest Path topic and added a bunch of content. Please review carefully.

### CODE REVIEW

File: docs/graphs/graph-queries/k-shortest-paths/index.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+4. York
+5. **Carlisle**
+6. **Birmingham**
+7. London
+
+The weight of this path is: 1.5 + 1.5 + 3.5 + 2.0 + 1.5 = **10.0**.
+
+#### Path via Glasgow
+
+Another route goes via Glasgow. This path has seven stations, but it's quicker if the edge weights are compared:
```
Comment (jamesm-macrometa) [inline docs/graphs/graph-queries/k-shortest-paths/index.md]:
```suggestion
Another route goes via Glasgow. This path has seven stations, but it's quicker if you compare the edge weights:
```

File: docs/graphs/graph-queries/k-shortest-paths/index.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+- **Portfolio Diversification**: k shortest path queries can analyze relationships between different investment assets to identify alternative paths for diversifying a portfolio, balancing risk and return.
+- **Fraud Detection**: k shortest path queries can help identify unusual patterns in financial transactions, which may indicate potential fraud. By considering different paths between entities, investigators can better understand the relationships between suspicious transactions and trace the flow of funds.
+
+## K Shortest Path Query in Action
+
+In this high-level example, there is a graph representing train stations and connections between cities. Each ellipse stands for a train station with the city name written inside of it (the vertices of the graph). Arrows represent train connections between cities (the edges of the graph), and the numbers near the arrows indicate the travel time between stations (used as edge weights).
```
Comment (jamesm-macrometa) [inline docs/graphs/graph-queries/k-shortest-paths/index.md]:
```suggestion
In this high-level example, a graph represents train stations and connections between cities. Each ellipse stands for a train station with the city name included in the station name (the vertices of the graph). Arrows represent train connections between cities (the edges of the graph), and the numbers near the arrows indicate the travel time between stations (used as edge weights).
```

**Review by jamesm-macrometa — APPROVED** [review]
LGTM!

File: docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+### Comparing SHORTEST_PATH and K_SHORTEST_PATHS with LIMIT 1
+
+The a route from **Aberdeen** to **London** and compare the outputs of `SHORTEST_PATH` and `K_SHORTEST_PATHS` with LIMIT 1. While both should return a path with the same length (or weight), they might not return the same path.
+
+Using `SHORTEST_PATH`:
+
+```sql
+FOR v, e IN OUTBOUND SHORTEST_PATH 'places/Aberdeen' TO 'places/London'
+GRAPH 'kShortestPathsGraph'
+    RETURN { place: v.label, travelTime: e.travelTime }
```
Comment (koshy-thomas) [inline docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md:105]:
```suggestion
    RETURN { place: v._key, travelTime: e.travelTime }
```

File: docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+![Train Connection Map](/img/train_map.png)
+
+### Comparing SHORTEST_PATH and K_SHORTEST_PATHS with LIMIT 1
+
+The a route from **Aberdeen** to **London** and compare the outputs of `SHORTEST_PATH` and `K_SHORTEST_PATHS` with LIMIT 1. While both should return a path with the same length (or weight), they might not return the same path.
+
+Using `SHORTEST_PATH`:
+
+```sql
+FOR v, e IN OUTBOUND SHORTEST_PATH 'places/Aberdeen' TO 'places/London'
```
Comment (koshy-thomas) [inline docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md:103]:
```suggestion
FOR v, e IN INBOUND SHORTEST_PATH 'places/Aberdeen' TO 'places/London'
```
There is no outbound path from aberdeen to london in this example

File: docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+
+## K Shortest Path Examples
+
+In this section, you'll find examples of K Shortest Paths queries using a named graph representing train connections in Europe.
+
+![Train Connection Map](/img/train_map.png)
+
+### Comparing SHORTEST_PATH and K_SHORTEST_PATHS with LIMIT 1
+
+The a route from **Aberdeen** to **London** and compare the outputs of `SHORTEST_PATH` and `K_SHORTEST_PATHS` with LIMIT 1. While both should return a path with the same length (or weight), they might not return the same path.
```
Comment (koshy-thomas) [inline docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md]: The start of the sentence isn't correct "The a route" ?

File: docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+
+```json
+[
+  {
+    "place": "Aberdeen",
+    "travelTime": null
+  },
+  {
+    "place": "Leuchars",
+    "travelTime": 1.5
```
Comment (koshy-thomas) [inline docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md:118]:
```suggestion
    "travelTime": 1
```

File: docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+  },
+  {
+    "place": "Edinburgh",
+    "travelTime": 1.5
```
Comment (koshy-thomas) [inline docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md:122]:
```suggestion
    "travelTime": 1.5
```
```suggestion
    "travelTime": 3
```

File: docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+  },
+  {
+    "place": "York",
+    "travelTime": 3.5
+  },
+  {
+    "place": "London",
+    "travelTime": 1.8
```
Comment (koshy-thomas) [inline docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md:130]:
```suggestion
    "travelTime": 2
```

File: docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+  }
+]
+```
+
+Using `K_SHORTEST_PATHS`:
+
+```sql
+FOR p IN OUTBOUND K_SHORTEST_PATHS 'places/Aberdeen' TO 'places/London'
```
Comment (koshy-thomas) [inline docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md:138]:
```suggestion
FOR p IN INBOUND K_SHORTEST_PATHS 'places/Aberdeen' TO 'places/London'
```
No outbound exists between aberdeen to London according to this example

File: docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+GRAPH 'kShortestPathsGraph'
+    LIMIT 1
+    RETURN { places: p.vertices[*].label, travelTimes: p.edges[*].travelTime }
```
Comment (koshy-thomas) [inline docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md:141]:
```suggestion
    RETURN { places: p.vertices[*]._key, travelTimes: p.edges[*].travelTime }
```

File: docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+  {
+    "places": [
+      "Aberdeen",
+      "Leuchars",
+      "Edinburgh",
+      "York",
+      "London"
+    ],
+    "travelTimes": [
+      1.5,
```
Comment (koshy-thomas) [inline docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md:157]:
```suggestion
      1,
```

File: docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+      1.5,
```
Comment (koshy-thomas) [inline docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md:158]:
```suggestion
      3,
```

File: docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md [hunk 13]
```diff
@@ -0,0 +1,357 @@
...
+      3.5,
```
Comment (koshy-thomas) [inline docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md:159]:
```suggestion
      4,
```

File: docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md [hunk 14]
```diff
@@ -0,0 +1,357 @@
...
+      1.8
```
Comment (koshy-thomas) [inline docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md:160]:
```suggestion
      2
```

File: docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md [hunk 15]
```diff
@@ -0,0 +1,357 @@
...
+  }
+]
+```
+
+### Finding Multiple Route Options
+
+This example demonstrates how to find more than one route option from **Aberdeen** to **London** using `K_SHORTEST_PATHS`:
+
+```sql
+FOR p IN OUTBOUND K_SHORTEST_PATHS 'places/Aberdeen' TO 'places/London'
```
Comment (koshy-thomas) [inline docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md:171]:
```suggestion
FOR p IN INBOUND K_SHORTEST_PATHS 'places/Aberdeen' TO 'places/London'
```

File: docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md [hunk 16]
```diff
@@ -0,0 +1,357 @@
...
+GRAPH 'kShortestPathsGraph'
+    LIMIT 3
+    RETURN {
+        places: p.vertices[*].label,
```
Comment (koshy-thomas) [inline docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md:175]:
```suggestion
        places: p.vertices[*]._key,
```

File: docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md [hunk 17]
```diff
@@ -0,0 +1,357 @@
...
+    "places": [
+      "Aberdeen",
+      "Leuchars",
+      "Edinburgh",
+      "York",
+      "London"
+    ],
+    "travelTimes": [
+      1.5,
+      1.5,
```
Comment (koshy-thomas) [inline docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md:195]:
```suggestion
      3,
```

File: docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md [hunk 18]
```diff
@@ -0,0 +1,357 @@
...
+      3.5,
```
Comment (koshy-thomas) [inline docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md:196]:
```suggestion
      4,
```

File: docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md [hunk 19]
```diff
@@ -0,0 +1,357 @@
...
+      1.8
+    ],
+    "travelTimeTotal": 8.3
```
Comment (koshy-thomas) [inline docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md:199]:
```suggestion
    "travelTimeTotal": 10
```

File: docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md (hunk 16 above)
Comment (koshy-thomas) [inline docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md]:
```suggestion
        places: p.vertices[*]._key,
```

File: docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md (hunk 10 above)
Comment (koshy-thomas) [inline docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md]:
```suggestion
    RETURN { places: p.vertices[*]._key, travelTimes: p.edges[*].travelTime }
```

File: docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md (hunk 3 above)
Comment (koshy-thomas) [inline docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md]:
```suggestion
    RETURN { place: v._key, travelTime: e.travelTime }
```

File: docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md [hunk 20]
```diff
@@ -0,0 +1,363 @@
...
+```sql
+FOR p IN OUTBOUND K_SHORTEST_PATHS 'places/StAndrews' TO 'places/Cologne'
+GRAPH 'kShortestPathsGraph'
+OPTIONS {
+    weightAttribute: 'travelTime',
+    defaultWeight: 15
+}
+    LIMIT 3
+    RETURN {
+        places: p.vertices[*].label,
```
Comment (koshy-thomas) [inline docs/graphs/graph-queries/k-shortest-paths/k-shortest-paths-examples.md]:
```suggestion
        places: p.vertices[*]._key,
```

**Comment by koshy-thomas** [comment]
@diana-macrometa I have corrected the sample data for connections edge collection. And corrected the query to use _key for city name instead of label (this attribute doesn't exist). You can truncate the connections collection and test the examples again. It should work. Let me know if you face any problems. Also check the comments above

**Comment by diana-macrometa** [comment]
Thanks so much, @koshy-thomas! Everything is working now.

**Review by koshy-thomas — APPROVED** [review]

**Review by martin-macrometa — APPROVED** [review]

**Pull request merged**